cli/lib/kontena/cli/services/exec_command.rb

```diff
@@ -12,14 +13,54 @@ class ExecCommand < Kontena::Command
     option ["-i", "--instance"], "INSTANCE", "Exec on given numbered instance, default first running" do |value| Integer(value) end
     option ["-a", "--all"], :flag, "Exec on all running instances"
     option ["--shell"], :flag, "Execute as a shell command"
+    option ["--interactive"], :flag, "Keep stdin open"
     option ["--skip"], :flag, "Skip failed instances when executing --all"
     option ["--silent"], :flag, "Do not show exec status"
     option ["--verbose"], :flag, "Show exec status"
 
     requires_current_master
     requires_current_grid
 
+    def execute
+      service_containers = client.get("services/#{parse_service_id(name)}/containers")['containers']
+      service_containers.sort_by! { |container| container['instance_number'] }
+      running_containers = service_containers.select{|container| container['status'] == 'running' }
+
+      if running_containers.empty?
+        exit_with_error "Service #{name} does not have any running containers"
+      end
+
+      if all?
+        ret = true
+        service_containers.each do |container|
+          if container['status'] == 'running'
+            if !exec_container(container)
+              ret = false
+            end
+          else
+            warning "Service #{name} container #{container['name']} is #{container['status']}, skipping"
+          end
+        end
+        return ret
+      elsif instance
+        if !(container = service_containers.find{|c| c['instance_number'] == instance})
+          exit_with_error "Service #{name} does not have container instance #{instance}"
+        elsif container['status'] != 'running'
+          exit_with_error "Service #{name} container #{container['name']} is not running, it is #{container['status']}"
+        else
+          interactive_exec(container)
+        end
+      else
+        if interactive?
+          interactive_exec(running_containers.first)
+        else 
+          exec_container(running_containers.first)
```

Comment:
@SpComb should we change "normal" exec to use websocket (because exec through http call has many problems)?